python/ray/air/_internal/torch_utils.py

```diff
@@ -128,9 +128,25 @@ def convert_ndarray_to_torch_tensor(
         ndarray: A NumPy ndarray that we wish to convert to a Torch Tensor.
         dtype: A Torch dtype for the created tensor; if None, the dtype will be
             inferred from the NumPy ndarray data.
+        device: The device on which the tensor should be placed; if None, the Torch
+            tensor will be constructed on the default device (automatic if running
+            inside a Ray Train worker, CPU otherwise).
 
     Returns: A Torch Tensor.
     """
+    if device is None:
+        # If we are in a TorchTrainer worker function, automatically
+        # use the correct device. This is tested in Train tests.
+
+        # Lazy import to avoid circular imports.
+        from ray.train.torch.train_loop_utils import get_device
+
+        try:
+            device = get_device()
+        except Exception:
+            # We are not running in a TorchTrainer.
+            pass
```

Comment:
Can this logic live at a higher layer and pass in the `device` param as needed? Seems a little magical to be done in every call to this utility.

Reply:
It can live either here or in the `Dataset`/`DatasetIterator` classes. Not sure where else we can fit it in

Reply:
+1.

I actually had a WIP PR from yesterday here: https://github.com/ray-project/ray/pull/31753